### PR TITLE
Solving problems with more topological dimensions

### PR DESCRIPTION
**Pull request by vlukes**
This PR  solves the problem of multiple topological dimension in one simulations, like `examples/acoustics/vibro_acoustic3d.py`, where 3D domains are coupled by a 2D interface.

The code changes in this  PR do the following:

- define region, which has its own `cmesh` with the topological dimension lower than other regions
- `conn_type` is  now a tuple: `(integration, tdim)`
-  setup up of  `dof_conn_types` and `geometry_types` is updated
- `Fields.get_connectivity()` is removed - alias `Field.get_econn()`
- unused  `v_tg` and `st_tg` attributes of `ConnInfo` are removed
- terms can have different DOF connectivity types, new list `dof_conn_types` replaces `dof_conn_type` in `ConnInfo`
- the number of components of vector fields is now determined by `region.cmesh.tdim`
- `examples/acoustics/vibro_acoustic3d.py` example does not need `ls.cm_pb` solver anymore, the solver may be removed in the future.

### CODE REVIEW

File: sfepy/discrete/fem/mappings.py [hunk 1]
```diff
@@ -11,9 +11,52 @@
...
+        bv = nm.cross(coors[:, 1, :] - coors[:, 0, :],
+                      coors[:, -1, :] - coors[:, 0, :])
+
+        b = nm.linalg.norm(bv, axis=1)
+        ab = nm.sqrt(bv[:, 0]**2 + bv[:, 1]**2)
+        cphi = bv[:, 2] / b
+        sphi = ab / b
+
+        mtxR = nm.eye(3, 3) * cphi[:, None, None]
+        idxs = nm.where(nm.abs(ab) > 1e-16)[0]
```
Comment (rc) [inline sfepy/discrete/fem/mappings.py:32]: `sfepy.linalg.geometry.make_axis_rotation_matrix()` ?

File: sfepy/examples/acoustics/vibro_acoustic3d.py [hunk 2]
```diff
@@ -104,42 +103,56 @@
...
+      = 0""" % (w2, ),
     }
 
 solvers = {
-    'ls': ('ls.cm_pb',
-           {'others': [base_dir
-                       + '/examples/acoustics/vibro_acoustic3d_mid.py'],
-            'coupling_variables': ['g0', 'w'],
-            }),
+    'ls': ('ls.scipy_direct', {}),
```
Comment (rc) [inline sfepy/examples/acoustics/vibro_acoustic3d.py:152]: auto_direct?

**Comment by rc** [comment]
Thanks! I have just skimmed over the changes, will look more soon. Good you got rid of the historic code! I have forgotten what `v_tg` and `ps_tg` did, and found those had been introduced in 2009, and have not been used recently. Still not sure what those did :)

File: sfepy/discrete/common/extmods/meshutils.c
Comment (rc) [inline sfepy/discrete/common/extmods/meshutils.c:1]: Does it change the original functionality, or is it mostly merging the 1,2,3D code, besides the introduction of `tdim`?

File: sfepy/discrete/equations.py [hunk 3]
```diff
@@ -393,18 +393,15 @@ def get_graph_conns(self, any_dof_conn=False, rdcs=None, cdcs=None,
...
-                continue
-
             rreg_name = info.get_region_name(can_trace=False)
             creg_name = info.get_region_name()
             mreg_name = None if creg_name == rreg_name else rreg_name
 
             rname = rvar.get_primary_name()
-            rkey = (rname, rreg_name, dct, None)
-            ckey = (cvar.name, creg_name, dct, mreg_name)
+            cpname = cvar.get_primary_name()
```
Comment (rc) [inline sfepy/discrete/equations.py:401]: cname?

File: sfepy/discrete/common/domain.py [hunk 4]
```diff
@@ -216,6 +264,11 @@ def create_region(self, name, select, kind='cell', parent=None,
         region = visit_stack(stack, region_op,
                              region_leaf(self, self.regions, select,
                                          functions, tdim))
+
+        if extra_options is not None and 'mesh_dim' in extra_options:
```
Comment (rc) [inline sfepy/discrete/common/domain.py:268]: Maybe use () to make the condition clearer? (`if (extra_options is not None) and ('mesh_dim' in extra_options):`)

File: sfepy/discrete/fem/domain.py [hunk 5]
```diff
@@ -105,17 +105,20 @@ def get_diameter(self):
...
-        for key, gel in self.geom_els.items():
+        if geom_els is None:
+            geom_els = self.geom_els
+
+        for key, gel in geom_els.items():
             ori = gel.orientation
 
             cmesh = self.cmesh_tdim[gel.dim]
-            if cmesh.tdim != cmesh.dim:
+            if cmesh.tdim != cmesh.dim and not force_check:
```
Comment (rc) [inline sfepy/discrete/fem/domain.py:121]: () and ()? I am nervous about implicit operator precedence :)

File: sfepy/discrete/fem/fields_base.py [hunk 6]
```diff
@@ -1071,21 +1071,38 @@ def get_coor(self, nods=None):
...
-
     def get_econn(self, conn_type, region, trace_region=None, local=False):
         """
         Get extended connectivity of the given type in the given region.
+
+        Parameters
+        ----------
+        conn_type: tuple or string
+            DOF connectivity type, eg. ('cell', 3) or 'cell'.
+            If the topological dimension not specified, taken from region.tdim.
```
Comment (rc) [inline sfepy/discrete/fem/fields_base.py:1082]: ... it is taken ...

File: sfepy/discrete/fem/fields_base.py [hunk 7]
```diff
@@ -1071,21 +1071,38 @@ def get_coor(self, nods=None):
...
+        region: sfepy.discrete.common.region.Region
+            The region for which the connectivity is required.
+        trace_region: None or string
+            If not None, return mirror connectivity according to `local`.
+        local: bool
+            If True, return local connectivity w.r.t. surface nodes,
```
Comment (rc) [inline sfepy/discrete/fem/fields_base.py:1088]: surface -> facet?

File: sfepy/discrete/fem/fields_base.py [hunk 8]
```diff
@@ -1071,21 +1071,38 @@ def get_coor(self, nods=None):
...
+            otherwise return global connectivity w.r.t. all mesh nodes.
+
+        Returns
+        -------
+        econn: numpy.ndarray
+            Connectivity information.
```
Comment (rc) [inline sfepy/discrete/fem/fields_base.py:1094]: The extended connectivity array.

econn = extended conn = connectivity with higher order nodes - mention this somewhere?

File: sfepy/examples/acoustics/vibro_acoustic3d.py (hunk 2 above)
Comment (rc) [inline sfepy/examples/acoustics/vibro_acoustic3d.py:152]: `ls.auto_direct`?

File: sfepy/examples/acoustics/vibro_acoustic3d.py
Comment (rc) [inline sfepy/examples/acoustics/vibro_acoustic3d.py:1]: It might be nice to put the definitions into a `define()` with `sound_speed` etc. as arguments.

**Comment by rc** [comment]
I have some comments/question (see above), but overall it looks very good!

Does it solve also #1003?

**Comment by rc** [comment]
IGA examples, e.g.
```
sfepy-run sfepy/examples/diffusion/poisson_iga.py
```
fail with:
```
 File "~/projects/sfepy-git/sfepy/discrete/iga/fields.py", line 247, in setup_extra_data
    dct, _ = info.dof_conn_type
AttributeError: 'ConnInfo' object has no attribute 'dof_conn_type'
```

File: sfepy/discrete/fem/mappings.py (hunk 1 above)
Comment (vlukes) [reply sfepy/discrete/fem/mappings.py:32]: I need to check this part of the code once again and I should create some tests.

File: sfepy/examples/acoustics/vibro_acoustic3d.py (hunk 2 above)
Comment (vlukes) [reply sfepy/examples/acoustics/vibro_acoustic3d.py:152]: Yes, `ls.auto_direct`.

File: sfepy/discrete/common/domain.py [hunk 9]
```diff
@@ -191,6 +191,54 @@ def reset_regions(self):
...
+        from sfepy.discrete.fem.geometry_element import (GeometryElement,
+            create_geometry_elements)
+        from sfepy.discrete import PolySpace
+        """
+        Create a new region which has its own cmesh with
+        topological dimension tdim.
+        """
+        mesh = self.mesh
+        if mesh.cmesh_tdim[tdim] is not None:
+            raise ValueError(f'cmesh of dimension {tdim} already exists!')
```
[the remark this answers is not in the document]
Comment (vlukes) [reply sfepy/discrete/common/domain.py:204]: There is a slot in `mesh.cmesh_tdim` list  for only  one cmesh of a given topological dimension. So something is probably  wrong if exists another cmesh with the same dimension.

File: sfepy/discrete/equations.py (hunk 3 above)
Comment (vlukes) [reply sfepy/discrete/equations.py:401]: Yes. Now, I can not see the reason why I choose `cpname`.

File: sfepy/discrete/fem/fields_base.py (hunk 8 above)
Comment (vlukes) [reply sfepy/discrete/fem/fields_base.py:1094]: > econn = extended conn = connectivity with higher order nodes - mention this somewhere?

Any idea where?

File: sfepy/examples/acoustics/vibro_acoustic3d.py
Comment (vlukes) [reply sfepy/examples/acoustics/vibro_acoustic3d.py:1]: Done

**Comment by vlukes** [comment]
> IGA examples, e.g.
> 
> ```
> sfepy-run sfepy/examples/diffusion/poisson_iga.py
> ```
> 
> fail with:
> 
> ```
>  File "~/projects/sfepy-git/sfepy/discrete/iga/fields.py", line 247, in setup_extra_data
>     dct, _ = info.dof_conn_type
> AttributeError: 'ConnInfo' object has no attribute 'dof_conn_type'
> ```

Yes, there was a bug in setup_extra_data(). Fixed.

**Comment by vlukes** [comment]
> Does it solve also #1003?

It would be nice, but, unfortunately, it does not solve that issue :(

**Comment by rc** [comment]
Thanks! The new changes fix the IGA issues. LGTM! Is it ready from your side?

**Comment by vlukes** [comment]
Yes, this PR is ready to be merged now.

26. října 2023 17:58:19 SELČ, Robert Cimrman ***@***.***> napsal:

**Pull request merged**